[ENG-1835] docs: update contributor docs (#1539)
amp-labs/connectors: CONTRIBUTING.md

# Contributing to Connectors

## Prerequisites

1. To test connectors, you'll need your own developer account and instance with the provider you're testing (like Salesforce or Hubspot).  Follow the provider guide on how to do this [here.](https://docs.withampersand.com/provider-guides/overview)

> 2. You'll need to obtain the credentials (using oAuth2.0 or apiKey) to link the connectors to the providers. 

Reviewer: ```suggestion
2. You'll need to obtain the credentials (using OAuth 2.0 or an api key) to link the connectors to the providers. 
```

> 3. Add a `creds.json` at the root of the connectors repo. Since each provider has different credential parameters and everyone has different environments, we've made a `CredentialRegistry` and `Reader` interface to make credential loading consistent. Please use these. We currently support `JSONReader`, `EnvVarReader`, and `ValueReader`.

Reviewer: ```suggestion
3. Add a `creds.json` file at the root of the `connectors` repo. Since each provider has different credential parameters and different environments, we've made a `CredentialRegistry` and `Reader` interface to make credential loading consistent which we recommend using for an easier development experience. We currently support `JSONReader`, `EnvVarReader`, and `ValueReader`.
```

## Adding a Proxy Connector 

1. Add a new file `connectors/providers/<PROVIDER>.go` for eg: `connectors/providers/linkedIn.go` (camelcase always!)
2. Update the file depending on the connector's auth type. 

### Provider Authentication Types

Here's a reference table for implementing different authentication types in your provider.go file:

| Auth Type | Description | Key Configuration Fields | Example Provider |
|-----------|-------------|-------------------------|------------------|
| OAuth2 Authorization Code (3-legged) | OAuth flow requiring user authorization | ```AuthType: OAuth2, OauthOpts: { GrantType: AuthorizationCode, AuthURL, TokenURL }``` | LinkedIn |
| OAuth2 Client Credentials (2-legged) | OAuth flow using client credentials | ```AuthType: OAuth2, OauthOpts: { GrantType: ClientCredentials, TokenURL }``` | Marketo |
| API Key | Authentication using an API key | ```AuthType: ApiKey, ApiKeyOpts: { Type: InHeader/InQuery, HeaderName, ValuePrefix }``` | SixSense |
| Basic Auth | Username/password authentication | ```AuthType: Basic, BaseURL``` | Insightly |

Additional configuration notes:

> - For workspace-specific providers, use `{{.workspace}}` in BaseURL

Reviewer: ```suggestion
- Use `{{.workspace}}` to denote workspace / account IDs for providers that require these to be a part of their API base URLs
```

> - Set `ExplicitScopesRequired: true` if the provider requires explicit scope definition

Reviewer: ```suggestion
- Set `ExplicitScopesRequired: true` if the provider requires scopes to be sent along with the oauth request
```

> - Use `PostAuthInfoNeeded: true` if additional information is needed after authentication

Reviewer: ```suggestion
- Use `PostAuthInfoNeeded: true` if additional metadata is needed after authentication for the connector to function (region, server, etc)
```

- Configure `Support` struct to specify provider capabilities (BulkWrite, Proxy, Read, Subscribe, Write)

Examples below for each Auth type: 

> ```go

Reviewer: Can we please link to the config files instead of having the contents here? My concern is that these might quickly get outdated, and we'll not remember to update it here

Author: @RajatPawar ah you're right, let me link this.

			HeaderName: "Authorization",
            ValuePrefix: "Token ",
            DocsURL: "https://api.6sense.com/docs/#get-your-api-token",
		},
        // For another provider, ValuePrefix may not be needed
        // For example, if the expected header is 'X-Api-Key: {your_api_key}'
        /*
        ApiKeyOpts: &ApiKeyOpts{
			HeaderName: "X-Api-Key",
            DocsURL: "https://api.6sense.com/docs/#get-your-api-token",
		}, */
		Support: Support{
			BulkWrite: BulkWriteSupport{
             Insert: false,
             Update: false,
             Upsert: false,
             Delete: false,
            },
			Proxy:     false,
			Read:      false,
			Subscribe: false,
			Write:     false,
		},
    },

    // Basic auth provider
    Insightly: {
        AuthType: Basic,
        BaseURL: "https://api.{{.pod}}.insightly.com",
		Support: Support {
			BulkWrite: BulkWriteSupport{
             Insert: false,
             Update: false,
             Upsert: false,
             Delete: false,
            },
			Proxy:     false,
			Read:      false,
			Subscribe: false,
			Write:     false,
		},
        PostAuthInfoNeeded: true,
    },
```

### Testing your proxy connector

> Proxy server script

Reviewer: ```suggestion
#### Proxy server script
```

`go run scripts/proxy/proxy.go` 

> This will run the proxy server at port `:4444` 
> 
> You can now make requests against `localhost:4444` and they will be routed to the provider you're working on.

Reviewer: ```suggestion
This will run the proxy server at port `:4444`. You can now make requests against `localhost:4444` and they will be routed to the provider you're working on.
```

> Note: For OAuth connectors, you'd need to run `go run scripts/oauth/token.go` to go through the OAuth flow first and get the credentials to save them into `creds.json`.

Once the proxy endpoint is tested with different kind of API calls, <b>you can make a PR. </b>

## Adding a Deep Connector

> Ensure you have made a proxy connector PR first and is merged and only then <b>work on the following changes in order</b>. 

Reviewer: ```suggestion
Ensure you have merged in a proxy connector before beginning <b>work on the following changes in order</b>. 
```

> Please make seperate PRs for each of the steps below. 
### 1. Add list metadata functionality: 

Files to add: 

1. For functionality add, `connectors/<PROVIDER>/metadata.go`.  
2. Implement the `SchemaProvider` interface in `connectors/<PROVIDER>/connector.go`
3. For testing add, `connectors/test/<PROVIDER>/metadata/metadata.go` 

Run your tests: 

`go run ./test/<PROVIDER>/metadata`

### 2. Add read action support for the objects agreed upon. 

Files to add or update: 
1. For functionality add, `connectors/<PROVIDER>/handlers.go`. Here we need to implement the interface `buildReadRequest` and `parseReadResponse`. 
2. Implement the `Reader` interface in `connectors/<PROVIDER>/connector.go`
3. For testing add, `connectors/<PROVIDER>/read.go`

> Run your tests: 

Reviewer: This feels quite repetitive, can we instead link to the `smartleadv2` connector and then ask contributors to refer to it?

Author: @RajatPawar I will add the `smarleadv2` connector reference here as well. Thanks! Although it wasn't very apparent to me which files were necessary for `read`  or `write` or `delete` from simply looking at the code, I'd say its ok to be repetitive in this case for clarity.
